#!/bin/bash

#########################################################
# 
# Platform: NCI Gadi HPC
# Description: extract indels from joint-genotyped VCF
# Author: Cali Willet
# cali.willet@sydney.edu.au
# Date last modified: 14/10/2020
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance 
# <or e.g. bioinformatics assistance of <PERSON> of Sydney Informatics
# Hub and resources and services from the National Computational 
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
# 
#########################################################

#PBS -P <project>
#PBS -N bsv-R1-S2
#PBS -l walltime=03:00:00
#PBS -l ncpus=1
#PBS -l mem=32GB
#PBS -l wd
#PBS -q hugemem
#PBS -W umask=022
#PBS -o ./Logs/bsv-R1-S2.o
#PBS -e ./Logs/bsv-R1-S2.e
#PBS -lstorage=<lstorage>

set -e

module load gatk/4.1.2.0 

round=<round>
step=2
type=INDEL

cohort=<cohort>
ref=<ref>
vcf_in=../GenotypeGVCFs/${cohort}.vcf
vcf_out=./VCFs/${cohort}_bootstrap-R${round}-S${step}_${type}_raw.vcf
log=./GATK_logs/${cohort}_bootstrap-R${round}-S${step}_${type}_raw.oe

rm -rf $log

echo "$(date): Bootstrap round ${round} step ${step}: Select ${type} variants" >> ${log} 2>&1
gatk --java-options "-Xmx29g -DGATK_STACKTRACE_ON_USER_EXCEPTION=true" \
	SelectVariants \
        -R $ref \
        -V $vcf_in \
        -select-type ${type} \
        -O $vcf_out >> ${log} 2>&1

echo "$(date): Finished." >> ${log} 2>&1

if grep -q -i error $log
then 
        printf "Error in GATK log ${log}\n" 
fi 

if grep -q Exception $log
then 
        printf "Exception in GATK log ${log}\n" 
fi
